In [1]:
import numpy as np
import pandas as pd


In [2]:
# import sys, os
# import pandas as pd
# import numpy as np

# from sklearn.linear_model import Lasso
# from sklearn.feature_selection import RFECV
# from sklearn.preprocessing import RobustScaler
# from sklearn.model_selection import StratifiedShuffleSplit, GridSearchCV
# from sklearn.metrics import mean_squared_error, mean_absolute_error, roc_auc_score, r2_score, make_scorer
# from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances

# def warn(*args, **kwargs):
#     pass
# import warnings
# warnings.warn = warn

# # some heuristic settings
# rfe_min_features = 12
# rfe_step = 15
# rfe_cv = 20
# sss_n_splits = 20
# sss_test_size = 0.35
# grid_search_cv = 20
# noise_std = 0.01
# r2_threshold = 0.185
# random_seed = 213

# np.random.seed(random_seed)

# # import data
# train = pd.read_csv('../input/dont-overfit-ii/train.csv')
# train_y = train['target']
# train_X = train.drop(['id','target'], axis=1).values

# test = pd.read_csv('../input/dont-overfit-ii/test.csv')
# test = test.drop(['id'], axis=1).values

# # scale using RobustScaler
# # fitting scaler on full data outperforms fitting on test_X only (+0.006 kaggle score)
# data = RobustScaler().fit_transform(np.concatenate((train_X, test), axis=0))
# train_X = data[:250]
# test = data[250:]

# # add a bit of noise to train_X to reduce overfitting
# train_X += np.random.normal(0, noise_std, train_X.shape)

# # define roc_auc_metric robust to only one class in y_pred
# def scoring_roc_auc(y, y_pred):
#     try:
#         return roc_auc_score(y, y_pred)
#     except:
#         return 0.5

# robust_roc_auc = make_scorer(scoring_roc_auc)

# # define model and its parameters
# model = Lasso(alpha=0.031, tol=0.01, random_state=random_seed, selection='random')

# param_grid = {
#             'alpha' : [0.022, 0.021, 0.02, 0.019, 0.023, 0.024, 0.025, 0.026, 0.027, 0.029, 0.031],
#             'tol'   : [0.0013, 0.0014, 0.001, 0.0015, 0.0011, 0.0012, 0.0016, 0.0017]
#         }

# # define recursive elimination feature selector
# feature_selector = RFECV(model, min_features_to_select=rfe_min_features, scoring=robust_roc_auc, step=rfe_step, verbose=0, cv=rfe_cv, n_jobs=-1)

# print("counter | val_mse  |  val_mae  |  val_roc  |  val_cos  |  val_dist  |  val_r2    | feature_count ")
# print("-------------------------------------------------------------------------------------------------")

# predictions = pd.DataFrame()
# counter = 0
# # split training data to build one model on each traing-data-subset
# for train_index, val_index in StratifiedShuffleSplit(n_splits=sss_n_splits, test_size=sss_test_size, random_state=random_seed).split(train_X, train_y):
#     X, val_X = train_X[train_index], train_X[val_index]
#     y, val_y = train_y[train_index], train_y[val_index]

#     # get the best features for this data set
#     feature_selector.fit(X, y)
#     # remove irrelevant features from X, val_X and test
#     X_important_features        = feature_selector.transform(X)
#     val_X_important_features    = feature_selector.transform(val_X)
#     test_important_features     = feature_selector.transform(test)

#     # run grid search to find the best Lasso parameters for this subset of training data and subset of features 
#     grid_search = GridSearchCV(feature_selector.estimator_, param_grid=param_grid, verbose=0, n_jobs=-1, scoring=robust_roc_auc, cv=20)
#     grid_search.fit(X_important_features, y)

#     # score our fitted model on validation data
#     val_y_pred = grid_search.best_estimator_.predict(val_X_important_features)
#     val_mse = mean_squared_error(val_y, val_y_pred)
#     val_mae = mean_absolute_error(val_y, val_y_pred)
#     val_roc = roc_auc_score(val_y, val_y_pred)
#     val_cos = cosine_similarity(val_y.values.reshape(1, -1), val_y_pred.reshape(1, -1))[0][0]
#     val_dst = euclidean_distances(val_y.values.reshape(1, -1), val_y_pred.reshape(1, -1))[0][0]
#     val_r2  = r2_score(val_y, val_y_pred)

#     # if model did well on validation, save its prediction on test data, using only important features
#     # r2_threshold (0.185) is a heuristic threshold for r2 error
#     # you can use any other metric/metric combination that works for you
#     if val_r2 > r2_threshold:
#         message = '<-- OK'
#         prediction = grid_search.best_estimator_.predict(test_important_features)
#         predictions = pd.concat([predictions, pd.DataFrame(prediction)], axis=1)
#     else:
#         message = '<-- skipping'


#     print("{0:2}      | {1:.4f}   |  {2:.4f}   |  {3:.4f}   |  {4:.4f}   |  {5:.4f}    |  {6:.4f}    |  {7:3}         {8}  ".format(counter, val_mse, val_mae, val_roc, val_cos, val_dst, val_r2, feature_selector.n_features_, message))
    
#     counter += 1

# print("-------------------------------------------------------------------------------------------------")
# print("{}/{} models passed validation threshold and will be ensembled.".format(len(predictions.columns), sss_n_splits))

# mean_pred = pd.DataFrame(predictions.mean(axis=1))
# mean_pred.index += 250
# mean_pred.columns = ['target']
# mean_pred.to_csv('submission_15.csv', index_label='id', index=True) 

In [3]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

## Data Exploration

In [4]:
train.head()

id  target      0      1      2      3      4      5      6      7  ...  \
0   0     1.0 -0.098  2.165  0.681 -0.614  1.309 -0.455 -0.236  0.276  ...   
1   1     0.0  1.081 -0.973 -0.383  0.326 -0.428  0.317  1.172  0.352  ...   
2   2     1.0 -0.523 -0.089 -0.348  0.148 -0.022  0.404 -0.023 -0.172  ...   
3   3     1.0  0.067 -0.021  0.392 -1.637 -0.446 -0.725 -1.035  0.834  ...   
4   4     1.0  2.347 -0.831  0.511 -0.021  1.225  1.594  0.585  1.509  ...   

     290    291    292    293    294    295    296    297    298    299  
0  0.867  1.347  0.504 -0.649  0.672 -2.097  1.051 -0.414  1.038 -1.065  
1 -0.165 -1.695 -1.257  1.359 -0.808 -1.624 -0.458 -1.099 -0.936  0.973  
2  0.013  0.263 -1.222  0.726  1.444 -1.165 -1.544  0.004  0.800 -1.211  
3 -0.404  0.640 -0.595 -0.966  0.900  0.467 -0.562 -0.254 -0.533  0.238  
4  0.898  0.134  2.415 -0.996 -1.006  1.378  1.246  1.478  0.428  0.253  

[5 rows x 302 columns]

In [5]:
print(train.shape, test.shape)

(250, 302) (19750, 301)


In [6]:
train.describe()

id      target           0           1           2           3  \
count  250.000000  250.000000  250.000000  250.000000  250.000000  250.000000   
mean   124.500000    0.640000    0.023292   -0.026872    0.167404    0.001904   
std     72.312977    0.480963    0.998354    1.009314    1.021709    1.011751   
min      0.000000    0.000000   -2.319000   -2.931000   -2.477000   -2.359000   
25%     62.250000    0.000000   -0.644750   -0.739750   -0.425250   -0.686500   
50%    124.500000    1.000000   -0.015500    0.057000    0.184000   -0.016500   
75%    186.750000    1.000000    0.677000    0.620750    0.805000    0.720000   
max    249.000000    1.000000    2.567000    2.419000    3.392000    2.771000   

                4           5           6           7  ...         290  \
count  250.000000  250.000000  250.000000  250.000000  ...  250.000000   
mean     0.001588   -0.007304    0.032052    0.078412  ...    0.044652   
std      1.035411    0.955700    1.006657    0.939731  ...    1.011416   
min     -2.566000   -2.845000   -2.976000   -3.444000  ...   -2.804000   
25%     -0.659000   -0.643750   -0.675000   -0.550750  ...   -0.617000   
50%     -0.023000    0.037500    0.060500    0.183500  ...    0.067500   
75%      0.735000    0.660500    0.783250    0.766250  ...    0.797250   
max      2.901000    2.793000    2.546000    2.846000  ...    2.865000   

              291         292         293         294         295         296  \
count  250.000000  250.000000  250.000000  250.000000  250.000000  250.000000   
mean     0.126344    0.018436   -0.012092   -0.065720   -0.106112    0.046472   
std      0.972567    0.954229    0.960630    1.057414    1.038389    0.967661   
min     -2.443000   -2.757000   -2.466000   -3.287000   -3.072000   -2.634000   
25%     -0.510500   -0.535750   -0.657000   -0.818500   -0.821000   -0.605500   
50%      0.091000    0.057500   -0.021000   -0.009000   -0.079500    0.009500   
75%      0.804250    0.631500    0.650250    0.739500    0.493000    0.683000   
max      2.801000    2.736000    2.596000    2.226000    3.131000    3.236000   

              297         298         299  
count  250.000000  250.000000  250.000000  
mean     0.006452    0.009372   -0.128952  
std      0.998984    1.008099    0.971219  
min     -2.776000   -3.211000   -3.500000  
25%     -0.751250   -0.550000   -0.754250  
50%      0.005500   -0.009000   -0.132500  
75%      0.794250    0.654250    0.503250  
max      2.626000    3.530000    2.771000  

[8 rows x 302 columns]

In [7]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Columns: 302 entries, id to 299
dtypes: float64(301), int64(1)
memory usage: 590.0 KB


 A classification model. With 299 features and 250 data points on the train. The test has 19750 data points. Values are not between 0 and 1, so might need some scaling and then PCA. Using a pipeline will be a good approach.
 
 Datatypes are mostly float and only one int.

In [8]:
# Lets check for missing values
print(f"Missing values in train sum: ",train.isnull().any().sum())
print(f"Missing values in test sum: ",test.isnull().any().sum())

Missing values in train sum:  0
Missing values in test sum:  0


In [9]:
print("Value counts of target in train set:")
train["target"].value_counts()


Value counts of target in train set:


1.0    160
0.0     90
Name: target, dtype: int64

Target value is not balanced. Accuracy metric should not be considered.

In [10]:
print("20 most correlated features to target variable:")
corr = train.corr().target.sort_values(ascending=False)
corr[:21]

20 most correlated features to target variable:


target    1.000000
33        0.373608
65        0.293846
24        0.173096
183       0.164146
199       0.159442
201       0.142238
30        0.132705
289       0.127213
114       0.124792
164       0.124151
101       0.118379
272       0.113909
226       0.113660
17        0.110998
105       0.110589
0         0.108966
244       0.108147
13        0.107828
176       0.099790
89        0.099526
Name: target, dtype: float64

## Baseline Model

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, precision_recall_curve,f1_score,confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report
seed = 42

In [12]:
test.head()

id      0      1      2      3      4      5      6      7      8  ...  \
0  250  0.500 -1.033 -1.595  0.309 -0.714  0.502  0.535 -0.129 -0.687  ...   
1  251  0.776  0.914 -0.494  1.347 -0.867  0.480  0.578 -0.313  0.203  ...   
2  252  1.750  0.509 -0.057  0.835 -0.476  1.428 -0.701 -2.009 -1.378  ...   
3  253 -0.556 -1.855 -0.682  0.578  1.592  0.512 -1.419  0.722  0.511  ...   
4  254  0.754 -0.245  1.173 -1.623  0.009  0.370  0.781 -1.763 -1.432  ...   

     290    291    292    293    294    295    296    297    298    299  
0 -0.088 -2.628 -0.845  2.078 -0.277  2.132  0.609 -0.104  0.312  0.979  
1 -0.683 -0.066  0.025  0.606 -0.353 -1.133 -3.138  0.281 -0.625 -0.761  
2 -0.094  0.351 -0.607 -0.737 -0.031  0.701  0.976  0.135 -1.327  2.463  
3 -0.336 -0.787  0.255 -0.031 -0.836  0.916  2.411  1.053 -1.601 -1.529  
4  2.184 -1.090  0.216  1.186 -0.143  0.322 -0.068 -0.156 -1.153  0.825  

[5 rows x 301 columns]

In [13]:

train_y = train['target']
train_x = train.drop(['id', 'target'], axis=1)

test = test.drop(['id'], axis=1)
scaler = StandardScaler()
X = train_x.copy()
y = train_y.copy()
X_scaled=scaler.fit_transform(X)
#train_df = X_scaled.copy()
y_df = y.copy()
test_scaled =scaler.transform(test)
# X_train_scaled,X_val_scaled,y_train_scaled,y_val_scaled = train_test_split(X_scaled,y, test_size=0.2,stratify=y)
# X_train,X_val,y_train,y_val = train_test_split(X,y, test_size=0.2,stratify=y)

In [14]:
logistic = LogisticRegression(random_state=seed)
pipe = Pipeline([
                ("scaler", StandardScaler()),
                ("model", logistic)   
                ])
pipe.fit(train_x,train_y)

Pipeline(steps=[('scaler', StandardScaler()),
                ('model', LogisticRegression(random_state=42))])

In [16]:
print(f"Train score:",pipe.score(train_x,train_y))
# print(f"Test score:",pipe.score(X_val, y_val))

Train score: 1.0


In [ ]:
# y_pred = pipe.predict(X_val)
# print(classification_report(y_val, y_pred, target_names=['0', '1']))

In [ ]:
# print(confusion_matrix(y_val, y_pred))

The model did poorly in classifying the model. On the leaderboard, it scored 0.51.

## Logistic Regression with L1 regularization of various strength

In [20]:
train_x = RobustScaler().fit_transform(train_x)
noise_std = 0.01
train_x += np.random.normal(0, noise_std, train_x.shape)
test = RobustScaler().fit_transform(test)

In [21]:
from sklearn.model_selection import cross_val_score
clf = LogisticRegression(class_weight='balanced', solver='liblinear', penalty ='l1', C= 0.1, max_iter=10000)
clf.fit(train_x, train_y)
print(f'5-fold val score : {cross_val_score(clf, train_x, train_y, cv=5)}')

5-fold val score : [0.72 0.7  0.76 0.74 0.7 ]


In [22]:
clf.fit(train_x, train_y)
result = clf.predict_proba(test)
result

array([[0.34743413, 0.65256587],
       [0.4918823 , 0.5081177 ],
       [0.50599509, 0.49400491],
       ...,
       [0.67504129, 0.32495871],
       [0.26886927, 0.73113073],
       [0.73147025, 0.26852975]])

In [ ]:
C = [10, 1, .1, .001]

for c in C:
    log_l1 = LogisticRegression(penalty='l1', C=c, solver='liblinear',random_state=seed)
    
    pipe = Pipeline([
                ("scaler", StandardScaler()),
                ("model", log_l1)])
    pipe.fit(X_train, y_train)
    print('C:', c)
    print('Training accuracy:', pipe.score(train_x, y_train))
    print('Test accuracy:', pipe.score(X_val, y_val))
    print('')

With an inverse of regularization strength C of 0.1 did better on both train and test set.

In [ ]:
log_l1_ = LogisticRegression(penalty='l1', C=1, solver='liblinear',random_state=seed)
pipe_log_l1 = Pipeline([
                ("scaler", StandardScaler()),
                ("model", log_l1_)   
                ])
pipe_log_l1.fit(X_train,y_train)
print(f"Train score:",pipe_log_l1.score(X_train, y_train))
print(f"Test score:",pipe_log_l1.score(X_val, y_val))
print("-------------------------------")
y_pred_log_l1 = pipe_log_l1.predict(X_val)
print(classification_report(y_val, y_pred_log_l1, target_names=['0', '1']))
print("-------------------------------")
print(confusion_matrix(y_val, y_pred_log_l1))

This submission didn't do well on the leader board. it scored 0.4890. With C = 1 improved the leaderboard score to 0.495. None has done better than the base model. Let's try other advanced classifiers like Random Forest, CatBoost, XGBoost with optimizations.

## One more thing - logistic regression with PCA

In [ ]:
# https://scikit-learn.org/stable/auto_examples/compose/plot_digits_pipe.html
# https://machinelearningmastery.com/multinomial-logistic-regression-with-python/
# Define a pipeline to search for the best combination of PCA truncation
# and classifier regularization.
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.ticker as ticker
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')


for i in range(10, 200, 10):
    pca = PCA(n_components=i)

    # set the tolerance to a large value to make the example faster
    logistic = LogisticRegression(solver='liblinear', tol=0.1, random_state=seed)
    pipe_log_pca = Pipeline([
                    ("scaler", StandardScaler()),
                    ("pca", pca),
                    ("model", logistic)   
                    ])
    pipe_log_pca.fit(X_train,y_train)
#     print(f"Train score:",pipe_log_pca.score(X_train, y_train))
#     print(f"Test score:",pipe_log_pca.score(X_val, y_val))
    print("-------------------------------")
    print('n-components:', i)
    print('Training accuracy:', pipe_log_pca.score(X_train, y_train))
    print('Test accuracy:', pipe_log_pca.score(X_val, y_val))
    print('')
#     y_pred_log_pca = pipe_log_pca.predict(X_val)
#     print(classification_report(y_val, y_pred_log_pca, target_names=['0', '1']))
#     print("-------------------------------")
#     print(confusion_matrix(y_val, y_pred_log_pca))
#     print("-------------------------------")
#     print(np.cumsum(pca.explained_variance_ratio_))

In [ ]:

pca = PCA(n_components=80)

    # set the tolerance to a large value to make the example faster
logistic = LogisticRegression(solver='liblinear', tol=0.1, random_state=seed)
pipe_log_pca = Pipeline([
                    ("scaler", StandardScaler()),
                    ("pca", pca),
                    ("model", logistic)   
                    ])
pipe_log_pca.fit(X_train,y_train)
print(f"Train score:",pipe_log_pca.score(X_train, y_train))
print(f"Test score:",pipe_log_pca.score(X_val, y_val))
print("-------------------------------")
#     print('n-components:', i)
#     print('Training accuracy:', pipe_log_pca.score(X_train, y_train))
#     print('Test accuracy:', pipe_log_pca.score(X_val, y_val))
#     print('')
y_pred_log_pca = pipe_log_pca.predict(X_val)
print(classification_report(y_val, y_pred_log_pca, target_names=['0', '1']))
print("-------------------------------")
print(confusion_matrix(y_val, y_pred_log_pca))
print("-------------------------------")
print(np.cumsum(pca.explained_variance_ratio_))

With PCA, obtained a score of .471 on the leaderboard.

## Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
forest = RandomForestClassifier(random_state = seed)
modelF = forest.fit(X_train_scaled, y_train_scaled)
y_predF = modelF.predict(X_val_scaled)
print('Training accuracy:', modelF.score(X_train_scaled, y_train_scaled))
print('Test accuracy:', modelF.score(X_val_scaled, y_val_scaled))

In [ ]:
n_estimators = [100, 300, 500, 800, 1200]
max_depth = [5, 8, 15, 25, 30]
min_samples_split = [2, 5, 10, 15, 100]
min_samples_leaf = [1, 2, 5, 10] 

hyperF = dict(n_estimators = n_estimators, max_depth = max_depth,  
              min_samples_split = min_samples_split, 
             min_samples_leaf = min_samples_leaf)

gridF = GridSearchCV(forest, hyperF, cv = 5, verbose = 1, 
                      n_jobs = -1)
bestF = gridF.fit(X_train_scaled, y_train_scaled)
bestF.best_score_

In [ ]:
bestF.best_params_

In [ ]:
forest_ = RandomForestClassifier(random_state = seed,max_depth= 5,min_samples_leaf= 1,min_samples_split=2,n_estimators=100)
modelF_ = forest_.fit(X_train_scaled, y_train_scaled)
y_predF_ = modelF_.predict(X_val_scaled)
print('Training accuracy:', modelF_.score(X_train_scaled, y_train_scaled))
print('Test accuracy:', modelF_.score(X_val_scaled, y_val_scaled))

Random forest with no fine tuning had a score of 0.499 on the leaderboard. With Gridsearchcv gives a score of 0.501.

## Trying out Loocv with Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV , train_test_split , cross_val_score
from sklearn.model_selection import LeaveOneOut
from sklearn.model_selection import LeavePOut
loocv = LeaveOneOut()
model_loocv = LogisticRegression(class_weight='balanced', C= 0.1, max_iter=10000).fit(X_scaled, y)
results_loocv = cross_val_score(model_loocv, X_scaled, y, cv=loocv)

In [ ]:
model_loocv.fit(X_scaled,y)

## Feature selection using RFECV on Lasso

In [ ]:
# define roc_auc_metric 
#then make it as ascorer 
from sklearn.preprocessing import RobustScaler
robust_scale_X = RobustScaler().fit_transform(X)
robust_scale_test = RobustScaler().fit_transform(X_test)
from sklearn.metrics import mean_squared_error, mean_absolute_error, roc_auc_score, r2_score, make_scorer
def scoring_roc_auc(y, y_pred):
    try:
        return roc_auc_score(y, y_pred)
    except:
        return 0.5

robust_roc_auc = make_scorer(scoring_roc_auc)

In [ ]:
from sklearn.linear_model import Lasso
from sklearn.feature_selection import RFECV
rfe_min_features = 12
rfe_step = 15
rfe_cv = 20
sss_n_splits = 20
sss_test_size = 0.35
grid_search_cv = 20
noise_std = 0.01
r2_threshold = 0.185

model = Lasso(alpha=0.031, tol=0.01, warm_start=True, random_state=seed, selection='random')
param_grid = {
    'alpha': [0.022,0.021,0.02, 0.019, 0.023, 0.024, 0.025, 0.026, 0.027, 0.029, 0.031],
    'tol': [0.0013, 0.0014, 0.001, 0.0015, 0.0011, 0.0012, 0.0016, 0.0017]}
# Define recursive elimination feature selector
feature_selector = RFECV(model, min_features_to_select=rfe_min_features, scoring=robust_roc_auc,
                        step=rfe_step, verbose=0, cv=rfe_cv, n_jobs=-1)

In [ ]:
print(train_df.shape)
print(y_df.shape)

In [ ]:
from sklearn.model_selection import StratifiedShuffleSplit, GridSearchCV
from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances
predictions = pd.DataFrame()
counter = 0
print("counter | val_mse  |  val_mae  |  val_roc  |  val_cos  |  val_dist  |  val_r2    | feature_count ")
print("-------------------------------------------------------------------------------------------------")
# split training data to build one model on each traing-data-subset
for train_index, val_index in StratifiedShuffleSplit(n_splits=sss_n_splits, test_size=sss_test_size, random_state=seed).split(X_scaled, y):
    X, val_X = robust_scale_X[train_index], robust_scale_X[val_index]
    y, val_y = y_df[train_index], y_df[val_index]

    # get the best features for this data set
    feature_selector.fit(X, y)
    # remove irrelevant features from X, val_X and test
    X_important_features        = feature_selector.transform(X)
    val_X_important_features    = feature_selector.transform(val_X)
    test_important_features     = feature_selector.transform(robust_scale_test)

    # run grid search to find the best Lasso parameters for this subset of training data and subset of features 
    grid_search = GridSearchCV(feature_selector.estimator_, param_grid=param_grid, verbose=0, n_jobs=-1, scoring=robust_roc_auc, cv=20)
    grid_search.fit(X_important_features, y)

    # score  fitted model on validation data
    val_y_pred = grid_search.best_estimator_.predict(val_X_important_features)
    val_mse = mean_squared_error(val_y, val_y_pred)
    val_mae = mean_absolute_error(val_y, val_y_pred)
    val_roc = roc_auc_score(val_y, val_y_pred)
    val_cos = cosine_similarity(val_y.reshape(1, -1), val_y_pred.reshape(1, -1))[0][0]
    val_dst = euclidean_distances(val_y.reshape(1, -1), val_y_pred.reshape(1, -1))[0][0]
    val_r2  = r2_score(val_y, val_y_pred)

    # if model did well on validation, save its prediction on test data, using only important features
    # r2_threshold (0.185) is a heuristic threshold for r2 error
    # you can use any other metric/metric combination that works for you
    if val_r2 > r2_threshold:
        message = '<-- OK'
        prediction = grid_search.best_estimator_.predict(test_important_features)
        predictions = pd.concat([predictions, pd.DataFrame(prediction)], axis=1)
    else:
        message = '<-- skipping'


    print("{0:2}      | {1:.4f}   |  {2:.4f}   |  {3:.4f}   |  {4:.4f}   |  {5:.4f}    |  {6:.4f}    |  {7:3}         {8}  ".format(counter, val_mse, val_mae, val_roc, val_cos, val_dst, val_r2, feature_selector.n_features_, message))
    
    counter += 1

print("-------------------------------------------------------------------------------------------------")
print("{}/{} models passed validation threshold and will be ensembled.".format(len(predictions.columns), sss_n_splits))

In [24]:
submission = pd.read_csv('sample_submission.csv')
submission['target'] = clf.predict_proba(test)
submission.to_csv('submission_31.csv', index=False)

submission.head()

id    target
0  250  0.347434
1  251  0.491882
2  252  0.505995
3  253  0.340229
4  254  0.536247

In [23]:
mean_pred = pd.DataFrame(predictions.mean(axis=1))
mean_pred.index += 250
mean_pred.columns = ['target']
mean_pred.to_csv('submission_31.csv', index_label='id', index=True) 

NameError: name 'predictions' is not defined